rllib/core/rl_trainer/trainer_runner.py

```diff
@@ -123,27 +165,40 @@ def _distributed_update(self, batch: MultiAgentBatch) -> List[Mapping[str, Any]]
             different backend than the cuda backend.
 
         Args:
-            batch: The data to use for the update.
+            See `.update()` docstring.
 
         Returns:
             A list of dictionaries of results from the updates from the RLTrainer(s)
         """
         refs = []
         global_size = len(self._workers)
-        batch_size = math.ceil(len(batch) / global_size)
         for i, worker in enumerate(self._workers):
             batch_to_send = {}
             for pid, sub_batch in batch.policy_batches.items():
                 batch_size = math.ceil(len(sub_batch) / global_size)
                 start = batch_size * i
                 end = min(start + batch_size, len(sub_batch))
                 batch_to_send[pid] = sub_batch[int(start) : int(end)]
+            # TODO (Avnish): int(batch_size) ? How should we shard MA batches really?
             new_batch = MultiAgentBatch(batch_to_send, int(batch_size))
-            refs.append(worker.update.remote(new_batch))
+            refs.append(
+                worker.update.remote(
+                    new_batch,
+                    minibatch_size=minibatch_size,
+                    num_iters=num_iters,
+                    reduce_fn=reduce_fn,
+                )
+            )
 
-        return ray.get(refs)
+        results = ray.get(refs)
+        return results
 
-    def additional_update(self, *args, **kwargs) -> List[Mapping[str, Any]]:
+    def additional_update(
+        self,
+        *,
+        reduce_fn: Optional[Callable[[ResultDict], ResultDict]] = _reduce_mean_results,
```

Comment:
since reduce_fn isnt passed to the actor but kwargs are, can we make users instead pass kwargs as an explicit dictionary?

hmm actually the more I think about it, why not make reduce_fn a member attribute of this class thats configured at initialization? I'm thinking about it and its not like users need to change this at every function call.

Reply:
The trainer-runner gets initialized at Algorithm initialization time, so it's not something user would normally configure. However, new algo devs, may need to implement their own reduce function and pass it at run-time, without touching the way that trainer_runner is initialized. in DQN you may need to not apply averaging to TD-errors and you may want to just concatenate them. The mental model is that I should create a custom reduce_fn for DQN and then pass it in at trainer_runner.update() call.

Reply:
gotcha

Reply:
still I think we should find a better way here to separate reduce_fn and kwargs since reduce_fn isn't getting routed. The simple choice would be to make people pass kwargs as a dict